File: leet/1008_bst_from_preorder_traversal/1008_bst_from_preorder_traversal.ts
// LeetCode 1008. Construct Binary Search Tree from Preorder Traversal
// https://leetcode.com/problems/construct-binary-search-tree-from-preorder-traversal/
// O(n) time, O(h) space
function bstFromPreorder(preorder: number[]): TreeNode | null {
    const root = new TreeNode(preorder[0]);
    let i = 1;
    const generateTree = (parent, minVal, maxVal) => {
        if (i >= preorder.length || preorder[i] > maxVal || preorder[i] < minVal)
            return;
        if (preorder[i] < parent.val) {
            const child = new TreeNode(preorder[i]);
            i++;
            parent.left = child;
            generateTree(child, minVal, parent.val);
        }
        
        if (i >= preorder.length || preorder[i] > maxVal || preorder[i] < minVal)
            return;
        if (preorder[i] > parent.val) {
            const child = new TreeNode(preorder[i]);
            i++;
            parent.right = child;
            generateTree(child, parent.val, maxVal);
        }
            
    };
    generateTree(root, Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
    return root;
};

File: leet/1008_bst_from_preorder_traversal/1008_bst_from_preorder_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> TreeNode:
        self.i = 1
        def insert(node, hi):
            if self.i < len(preorder) and preorder[self.i] < node.val:
                node.left = TreeNode(preorder[self.i])
                self.i += 1
                insert(node.left, node.val)
            if self.i < len(preorder) and preorder[self.i] <= hi:
                node.right = TreeNode(preorder[self.i])
                self.i += 1
                insert(node.right, hi)
            
        root = TreeNode(preorder[0])
        insert(root, float('inf'))
        return root
